docs/docs/noir/concepts/data_types/arrays.md

```diff
@@ -77,6 +77,42 @@ However, multidimensional slices are not supported. For example, the following c
 let slice : [[Field]] = &[];
 ```
 
+## Dynamic Indexing
+
+Using constant indices of arrays will often be more efficient at runtime in constrained code.
+Indexing an array with non-constant indices (indices derived from the inputs to the program) is also
```

Comment:
I was thinking of how to implement this in the fuzzer and came up with a scheme to propagate the information about which variable and expression can be originated back to _inputs to the program_. 

I tried this program, on the code that doesn't have this PR:
```rust
fn main() -> pub bool {
    let b = [&mut true, &mut false];
    let mut c = false;
    for k in 0..=1 {
        c |= *b[k];
    }
    c
}
```
It crashes the compiler, even though there simply no index derived from program inputs. 

Is there another way of phrasing this restriction that better reflects the rules? In practice I see it's anything that cannot be simplified to a constant by the compiler, which inputs certainly are, but they don't cover the whole story.

Reply:
This code above actually works with the fix in https://github.com/noir-lang/noir/pull/8877

Reply:
> Okay then I assume this phrasing is correct and if the compiler for any reason cannot figure out a numeric constant for an index not dependent on input parameters, then it's a bug.

Yeah, we _should_ guarantee references are removed from the program before acir. So the only way they can still be in is if they're used in some way we inherently cannot optimize out. In theory anything taking constant inputs only which does not have side-effects should be able to be optimized out. If there are more side-effectful places references can be used in they also may not be optimized out there but we already ban these (no references in println, oracles, acir->brillig calls)